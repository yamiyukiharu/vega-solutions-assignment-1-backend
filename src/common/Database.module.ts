import { Module } from "@nestjs/common";
import { TypeOrmModule } from "@nestjs/typeorm";
import { AppConfigModule } from "./AppConfig.module";
import { ConfigService } from "@nestjs/config";
import { Transaction } from "src/transaction/models/Transaction.model";

@Module({
  imports: [
    TypeOrmModule.forRootAsync({
      imports: [AppConfigModule],
      useFactory: (configService: ConfigService) => ({
        type: 'mongodb',
        url: configService.get('DB_URL'),
        entities: [Transaction],
        synchronize: true,
        useNewUrlParser: true,
        logging: true,
      }),
      inject: [ConfigService],
    }),
  ],
  exports: [TypeOrmModule]
})
export class DatabaseModule { }