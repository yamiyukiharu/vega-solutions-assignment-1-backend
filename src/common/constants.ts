export const REPORTS_QUEUE = 'transactions-report';

export const RECORD_QUEUE = 'transactions-record';