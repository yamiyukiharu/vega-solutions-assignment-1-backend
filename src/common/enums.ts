export enum Protocol {
  UNISWAPV3 = 'uniswapv3',
}

export enum Pool {
  ETH_USDC = 'usdc_eth',
}

export enum ReportStatus {
  PENDING = 'pending',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
  FAILED = 'failed',
}