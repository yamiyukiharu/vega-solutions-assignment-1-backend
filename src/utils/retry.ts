export const retryOnFail = async <T>(
  fn: () => Promise<T>,
  retryThreshold: number,
): Promise<T> => {
  let retryCount = 0;
  let error: Error | null = null;

  while (retryCount < retryThreshold) {
    try {
      const res = await fn();
      return res;
    } catch (e) {
      error = e;
    }
    retryCount++;
  }

  throw error;
};
