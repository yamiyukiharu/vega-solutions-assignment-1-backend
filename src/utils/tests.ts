export const mongoSnapshotIgnoreFields = {
  _id: expect.any(Object),
  createdAt: expect.any(Date),
  updatedAt: expect.any(Date),
};