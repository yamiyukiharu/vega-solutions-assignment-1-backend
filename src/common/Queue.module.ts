
import { BullModule } from '@nestjs/bull';
import { Module } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { AppConfigModule } from './AppConfig.module';

@Module({
  imports: [
    BullModule.registerQueueAsync({
      imports: [AppConfigModule],
      name: 'transactions-report',
      useFactory: async (configService: ConfigService) => ({
        redis: {
          host: configService.get('QUEUE_HOST'),
          port: configService.get('QUEUE_PORT'),
        },
      }),
      inject: [ConfigService],
    }),
  ],
  exports: [BullModule],
})
export class QueueModule { }
