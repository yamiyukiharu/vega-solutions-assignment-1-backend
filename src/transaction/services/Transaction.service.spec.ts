import { Pool, Protocol, ReportStatus } from 'src/common/enums';
import { Test, TestingModule } from '@nestjs/testing';
import { TransactionService } from './Transaction.service';
import { anything, instance, mock, reset, verify, when } from 'ts-mockito';
import { HttpService } from '@nestjs/axios';
import { IExchangeRateProvider } from 'src/exchange-rate/providers/IExchangeRate.provider';
import { ITransactionProvider } from '../providers/ITransaction.provider';

import { MongoMemoryServer } from 'mongodb-memory-server';
import { Connection, connect, Model } from 'mongoose';
import { getModelToken } from '@nestjs/mongoose';
import { getQueueToken } from '@nestjs/bull';
import { Transaction, TransactionSchema } from '../models/Transaction.model';
import {
  TransactionReport,
  TransactionReportSchema,
} from '../models/TransactionReport.model';
import { QueueModule } from 'src/common/modules/Queue.module';
import { ExchangeRateService } from 'src/exchange-rate/services/ExchangeRate.service';
import { EtherscanProvider } from '../providers/Etherscan.provider';
import { BinanceProvider } from 'src/exchange-rate/providers/Binance.provider';
import { mongoSnapshotIgnoreFields, sanitizeDocuments } from 'src/utils/tests';
import { REPORTS_QUEUE } from 'src/common/constants';
import { BullModule } from '@nestjs/bull';
import * as dayjs from 'dayjs';
import {
  RecordInterval,
  RecordIntervalSchema,
} from '../models/RecordInterval.model';
import { CACHE_MANAGER } from '@nestjs/cache-manager';

describe('TransactionService', () => {
  let service: TransactionService;
  let mongod: MongoMemoryServer;
  let mongoConnection: Connection;
  let transactionModel: Model<Transaction>;
  let reportModel: Model<TransactionReport>;
  let intervalModel: Model<RecordInterval>;
  let transactionProviderMock = mock(EtherscanProvider);
  let exchangeRateProviderMock = mock(BinanceProvider);

  const mockQueue: any = {
    add: jest.fn(),
    process: jest.fn(),
    on: jest.fn(),
  };

  beforeAll(async () => {
    // mock database
    mongod = await MongoMemoryServer.create();
    const uri = mongod.getUri();
    mongoConnection = (await connect(uri)).connection;
    transactionModel = mongoConnection.model(
      Transaction.name,
      TransactionSchema,
    );
    reportModel = mongoConnection.model(
      TransactionReport.name,
      TransactionReportSchema,
    );
    intervalModel = mongoConnection.model(
      RecordInterval.name,
      RecordIntervalSchema,
    );

    const module: TestingModule = await Test.createTestingModule({
      imports: [
        BullModule.registerQueue({
          name: REPORTS_QUEUE,
        }),
      ],
      providers: [
        TransactionService,
        ExchangeRateService,
        {
          provide: getModelToken(Transaction.name),
          useValue: transactionModel,
        },
        {
          provide: getModelToken(TransactionReport.name),
          useValue: reportModel,
        },
        {
          provide: getModelToken(RecordInterval.name),
          useValue: intervalModel,
        },
        {
          provide: HttpService,
          useValue: instance(mock(HttpService)),
        },
        {
          provide: ITransactionProvider,
          useValue: instance(transactionProviderMock),
        },
        {
          provide: IExchangeRateProvider,
          useValue: instance(exchangeRateProviderMock),
        },
        {
          provide: CACHE_MANAGER,
          useValue: {
            get: jest.fn(),
            set: jest.fn(),
          },
        },
      ],
    })
      // mock queue
      .overrideProvider(getQueueToken(REPORTS_QUEUE))
      .useValue(mockQueue)
      .compile();

    service = module.get<TransactionService>(TransactionService);
  });

  beforeEach(async () => {
    when(transactionProviderMock.getTransactions(anything())).thenResolve([
      {
        id: '0x1',
        timestamp: 1,
        blockNumber: 1,
        fee: '22921930374619520',
      },
      {
        id: '0x2',
        timestamp: 2,
        blockNumber: 2,
        fee: '14691703394596320',
      },
      {
        id: '0x3',
        timestamp: 5,
        blockNumber: 3,
        fee: '14691703394596320',
      },
    ]);

    when(exchangeRateProviderMock.getHistoricalRates(anything()))
      .thenResolve([
        { timestamp: 1, value: '1800' },
        { timestamp: 2, value: '1801' },
        { timestamp: 3, value: '1802' },
      ])
      .thenResolve([
        { timestamp: 4, value: '1803' },
        { timestamp: 5, value: '2004' },
        { timestamp: 6, value: '1805' },
      ]);
  });

  afterAll(async () => {
    await mongoConnection.dropDatabase();
    await mongoConnection.close();
    await mongod.stop();
  });

  afterEach(async () => {
    const collections = mongoConnection.collections;
    for (const key in collections) {
      const collection = collections[key];
      await collection.deleteMany({});
    }
    reset(transactionProviderMock);
  });

  describe('getTransactions', () => {
    it('should return transactions for a given protocol and pool', async () => {
      // Arrange
      const protocol = Protocol.UNISWAPV3;
      const pool = Pool.ETH_USDC;
      const doc = await transactionModel.create({
        protocol,
        pool,
        hash: '0x1',
        timestamp: 1,
        blockNumber: 1,
        fee: {
          eth: '100000000000000000',
          usdt: '100',
        },
      });

      // Act
      const result = await service.getTransactionByHash('0x1', protocol, pool);

      // Assert
      expect(result).toMatchSnapshot();
    });

    it('should return null if no transaction is found', async () => {
      // Arrange
      const protocol = Protocol.UNISWAPV3;
      const pool = Pool.ETH_USDC;

      // Act
      const result = await service.getTransactionByHash('0x1', protocol, pool);

      // Assert
      expect(result).toBeNull();
    });
  });

  describe('getTransactionsList', () => {
    it('should return transactions for a given protocol and pool', async () => {
      // Arrange
      const protocol = Protocol.UNISWAPV3;
      const pool = Pool.ETH_USDC;
      for (let i = 0; i < 10; i++) {
        await transactionModel.create({
          protocol,
          pool,
          hash: `0x${i}`,
          timestamp: i,
          blockNumber: i,
          fee: {
            eth: '100000000000000000',
            usdt: '100',
          },
        });
      }

      // Act
      const result = await service.getTransactionList(protocol, pool, 0, 5);

      // Assert
      expect(result).toMatchSnapshot();
    });
  });

  describe('recordNewTransactions', () => {
    it('can record new transactions', async () => {
      // Arrange
      const protocol = Protocol.UNISWAPV3;
      const pool = Pool.ETH_USDC;

      // Act
      await service.recordNewTransactions(protocol, pool);

      // Assert
      const data = await transactionModel.find({});
      const sanitizedData = sanitizeDocuments(data);
      expect(sanitizedData).toMatchSnapshot();
    });
  });
  describe('getTransactionCount', () => {
    it('should return the count of transactions for a given protocol and pool', async () => {
      // Arrange
      const protocol = Protocol.UNISWAPV3;
      const pool = Pool.ETH_USDC;
      const count = 10;

      for (let i = 0; i < count; i++) {
        await transactionModel.create({
          protocol,
          pool,
          hash: `0x${i}`,
          timestamp: 1,
          blockNumber: 1,
          fee: {
            eth: '100000000000000000',
            usdt: '100',
          },
        });
      }

      // Act
      const result = await service.getTransactionCount(protocol, pool);

      // Assert
      expect(result).toEqual(count);
    });
  });

  describe('triggerReportGeneration', () => {
    it('should create a report and add it to the queue', async () => {
      // Arrange
      const protocol = Protocol.UNISWAPV3;
      const pool = Pool.ETH_USDC;
      const startTime = new Date('2022-01-01T00:00:00.000Z');
      const endTime = new Date('2022-01-02T00:00:00.000Z');

      const result = await service.triggerReportGeneration(
        protocol,
        pool,
        startTime,
        endTime,
      );

      // Act
      const doc = await reportModel.findById(result);

      // Assert
      expect(doc.toObject()).toMatchSnapshot(mongoSnapshotIgnoreFields);
      expect(mockQueue.add).toHaveBeenCalledTimes(1);
      expect(mockQueue.add).toHaveBeenCalledWith({ id: result });
    });
  });

  describe('getReportStatus', () => {
    it('should return the status of a report', async () => {
      const reportId = '123';
      const report = {
        status: ReportStatus.PENDING,
      };
      jest.spyOn(reportModel, 'findById').mockResolvedValueOnce(report as any);

      const result = await service.getReportStatus(reportId);

      expect(reportModel.findById).toHaveBeenCalledWith(reportId);
      expect(result).toEqual(report.status);
    });

    it('should return null if the report is not found', async () => {
      const reportId = '123';
      jest.spyOn(reportModel, 'findById').mockResolvedValueOnce(null);

      const result = await service.getReportStatus(reportId);

      expect(reportModel.findById).toHaveBeenCalledWith(reportId);
      expect(result).toBeNull();
    });
  });

  describe('processReport', () => {
    it('should update report status to COMPLETED and set total fees', async () => {
      // Arrange
      const { id } = await reportModel.create({
        protocol: Protocol.UNISWAPV3,
        pool: Pool.ETH_USDC,
        startTimestamp: dayjs('2021-01-01T00:00:00.000Z').unix(),
        endTimestamp: dayjs('2021-01-02T00:00:00.000Z').unix(),
      });

      // Act
      await service.processReport(id);

      // Assert
      const res = await reportModel.findById(id);
      const transactions = await transactionModel.find({});
      const sanitizedData = sanitizeDocuments(transactions);

      expect(res.toObject()).toMatchSnapshot(mongoSnapshotIgnoreFields);
      expect(sanitizedData).toMatchSnapshot();
    });

    it('should not process if the data is already present in database', async () => {
      // Arrange
      await intervalModel.create({
        protocol: Protocol.UNISWAPV3,
        pool: Pool.ETH_USDC,
        start: dayjs('2021-01-01T00:00:00.000Z').unix(),
        end: dayjs('2021-01-02T00:00:00.000Z').unix(),
      });

      const { id } = await reportModel.create({
        protocol: Protocol.UNISWAPV3,
        pool: Pool.ETH_USDC,
        startTimestamp: dayjs('2021-01-01T00:00:00.000Z').unix(),
        endTimestamp: dayjs('2021-01-02T00:00:00.000Z').unix(),
      });

      // Act
      await service.processReport(id);

      // Assert
      const res = await reportModel.findById(id);
      const transactions = await transactionModel.find({});
      const sanitizedData = sanitizeDocuments(transactions);

      verify(transactionProviderMock.getTransactions(anything())).never();
      expect(res.toObject()).toMatchSnapshot(mongoSnapshotIgnoreFields);
      expect(sanitizedData).toMatchSnapshot();
    });

    it('should update report status to FAILED if an error occurs', async () => {
      // Arrange
      const { id } = await reportModel.create({
        protocol: Protocol.UNISWAPV3,
        pool: Pool.ETH_USDC,
        startTimestamp: dayjs('2021-01-01T00:00:00.000Z').unix(),
        endTimestamp: dayjs('2021-01-02T00:00:00.000Z').unix(),
      });

      when(transactionProviderMock.getTransactions(anything())).thenReject(
        new Error('test error'),
      );

      // Act
      await service.processReport(id);

      // Assert
      const res = await reportModel.findById(id);
      expect(res.toObject()).toMatchSnapshot(mongoSnapshotIgnoreFields);
    });
  });

  describe('getReport', () => {
    it('should return a report', async () => {
      // Arrange
      const { id } = await reportModel.create({
        protocol: Protocol.UNISWAPV3,
        pool: Pool.ETH_USDC,
        startTimestamp: 2,
        endTimestamp: 5,
        totalFee: {
          eth: '200000000000000000',
          usdt: '200',
        },
        count: 2,
        status: ReportStatus.COMPLETED,
      });

      await transactionModel.create([
        {
          protocol: Protocol.UNISWAPV3,
          pool: Pool.ETH_USDC,
          hash: '0x1',
          timestamp: 1,
          blockNumber: 1,
          fee: {
            eth: '100000000000000000',
            usdt: '100',
          },
        },
        {
          protocol: Protocol.UNISWAPV3,
          pool: Pool.ETH_USDC,
          hash: '0x2',
          timestamp: 2,
          blockNumber: 2,
          fee: {
            eth: '100000000000000000',
            usdt: '100',
          },
        },
        {
          protocol: Protocol.UNISWAPV3,
          pool: Pool.ETH_USDC,
          hash: '0x3',
          timestamp: 3,
          blockNumber: 3,
          fee: {
            eth: '100000000000000000',
            usdt: '100',
          },
        },
      ]);
      // Act
      const result = await service.getReport(id, 0, 10);

      // Assert
      result.id = undefined;
      expect(result).toMatchSnapshot();
    });
  });
});
