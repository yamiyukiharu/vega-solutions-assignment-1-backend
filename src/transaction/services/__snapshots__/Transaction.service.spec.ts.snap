// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionService getReport should return a report 1`] = `
{
  "data": [
    {
      "__v": 0,
      "blockNumber": 3,
      "fee": {
        "eth": "100000000000000000",
        "usdt": "100",
      },
      "hash": "0x3",
      "pool": "eth_usdc",
      "protocol": "uniswapv3",
      "timestamp": 3,
    },
    {
      "__v": 0,
      "blockNumber": 2,
      "fee": {
        "eth": "100000000000000000",
        "usdt": "100",
      },
      "hash": "0x2",
      "pool": "eth_usdc",
      "protocol": "uniswapv3",
      "timestamp": 2,
    },
  ],
  "limit": 10,
  "page": 0,
  "total": 2,
  "totalFee": {
    "eth": "200000000000000000",
    "usdt": "200",
  },
}
`;

exports[`TransactionService getTransactions should return transactions for a given protocol and pool 1`] = `
{
  "__v": 0,
  "blockNumber": 1,
  "fee": {
    "eth": "100000000000000000",
    "usdt": "100",
  },
  "hash": "0x1",
  "pool": "eth_usdc",
  "protocol": "uniswapv3",
  "timestamp": 1,
}
`;

exports[`TransactionService getTransactionsList should return transactions for a given protocol and pool 1`] = `
[
  {
    "__v": 0,
    "blockNumber": 9,
    "fee": {
      "eth": "100000000000000000",
      "usdt": "100",
    },
    "hash": "0x9",
    "pool": "eth_usdc",
    "protocol": "uniswapv3",
    "timestamp": 9,
  },
  {
    "__v": 0,
    "blockNumber": 8,
    "fee": {
      "eth": "100000000000000000",
      "usdt": "100",
    },
    "hash": "0x8",
    "pool": "eth_usdc",
    "protocol": "uniswapv3",
    "timestamp": 8,
  },
  {
    "__v": 0,
    "blockNumber": 7,
    "fee": {
      "eth": "100000000000000000",
      "usdt": "100",
    },
    "hash": "0x7",
    "pool": "eth_usdc",
    "protocol": "uniswapv3",
    "timestamp": 7,
  },
  {
    "__v": 0,
    "blockNumber": 6,
    "fee": {
      "eth": "100000000000000000",
      "usdt": "100",
    },
    "hash": "0x6",
    "pool": "eth_usdc",
    "protocol": "uniswapv3",
    "timestamp": 6,
  },
  {
    "__v": 0,
    "blockNumber": 5,
    "fee": {
      "eth": "100000000000000000",
      "usdt": "100",
    },
    "hash": "0x5",
    "pool": "eth_usdc",
    "protocol": "uniswapv3",
    "timestamp": 5,
  },
]
`;

exports[`TransactionService processReport should update report status to COMPLETED and set total fees 1`] = `
{
  "__v": Any<Number>,
  "_id": Any<Object>,
  "count": 3,
  "endTimestamp": 1609545600,
  "pool": "eth_usdc",
  "protocol": "uniswapv3",
  "startTimestamp": 1609459200,
  "status": "completed",
  "totalFee": {
    "eth": "52305337163812160",
    "usdt": "97.1614060907541336",
  },
}
`;

exports[`TransactionService processReport should update report status to COMPLETED and set total fees 2`] = `
[
  {
    "__v": undefined,
    "_id": undefined,
    "blockNumber": 1,
    "fee": {
      "eth": "22921930374619520",
      "usdt": "41.259474674315136",
    },
    "hash": "0x1",
    "pool": "eth_usdc",
    "protocol": "uniswapv3",
    "timestamp": 1,
  },
  {
    "__v": undefined,
    "_id": undefined,
    "blockNumber": 2,
    "fee": {
      "eth": "14691703394596320",
      "usdt": "26.45975781366797232",
    },
    "hash": "0x2",
    "pool": "eth_usdc",
    "protocol": "uniswapv3",
    "timestamp": 2,
  },
  {
    "__v": undefined,
    "_id": undefined,
    "blockNumber": 3,
    "fee": {
      "eth": "14691703394596320",
      "usdt": "29.44217360277102528",
    },
    "hash": "0x3",
    "pool": "eth_usdc",
    "protocol": "uniswapv3",
    "timestamp": 5,
  },
]
`;

exports[`TransactionService processReport should update report status to FAILED if an error occurs 1`] = `
{
  "__v": Any<Number>,
  "_id": Any<Object>,
  "endTimestamp": 1609545600,
  "pool": "eth_usdc",
  "protocol": "uniswapv3",
  "startTimestamp": 1609459200,
  "status": "failed",
}
`;

exports[`TransactionService recordNewTransactions can record new transactions 1`] = `
[
  {
    "__v": undefined,
    "_id": undefined,
    "blockNumber": 1,
    "fee": {
      "eth": "22921930374619520",
      "usdt": "41.259474674315136",
    },
    "hash": "0x1",
    "pool": "eth_usdc",
    "protocol": "uniswapv3",
    "timestamp": 1,
  },
  {
    "__v": undefined,
    "_id": undefined,
    "blockNumber": 2,
    "fee": {
      "eth": "14691703394596320",
      "usdt": "26.45975781366797232",
    },
    "hash": "0x2",
    "pool": "eth_usdc",
    "protocol": "uniswapv3",
    "timestamp": 2,
  },
  {
    "__v": undefined,
    "_id": undefined,
    "blockNumber": 3,
    "fee": {
      "eth": "14691703394596320",
      "usdt": "29.44217360277102528",
    },
    "hash": "0x3",
    "pool": "eth_usdc",
    "protocol": "uniswapv3",
    "timestamp": 5,
  },
]
`;

exports[`TransactionService triggerReportGeneration should create a report and add it to the queue 1`] = `
{
  "__v": Any<Number>,
  "_id": Any<Object>,
  "endTimestamp": 1641081600,
  "pool": "eth_usdc",
  "protocol": "uniswapv3",
  "startTimestamp": 1640995200,
  "status": "pending",
}
`;
